### PR TITLE
Swift: extract `MacroDecl`

### PR DESCRIPTION
**Pull request by AlexDenisov**
The other three elements are skipped (`MacroExpansionExpr`, `MacroExpansionDecl`, `MissingDecl`) as they only appear in cases when a macro is not expanded (due to an error).

### CODE REVIEW

File: swift/ql/lib/codeql/swift/elements/MacroRole.qll [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * This module provides a hand-modifiable wrapper around the generated class `MacroRole`.
+ */
+
+private import codeql.swift.generated.MacroRole
+
+class MacroRole extends Generated::MacroRole {
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/MacroRole.qll:7]: Could you add some QLDoc to describe what a `MacroRole` is? 🙏

It looks like you added it to the autogenerated class via the `schena.py` file (which is awesome!), but I think we should port it to the user-facing file as well.

File: swift/ql/test/library-tests/ast/PrintAst.expected [hunk 2]
```diff
@@ -1,3 +1,13 @@
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(expression)
+#-----| [MacroRole] #freestanding(declaration)
+#-----| [MacroRole] #freestanding(declaration)
```
Comment (MathiasVP) [inline swift/ql/test/library-tests/ast/PrintAst.expected:10]: It looks like we should add `MacroRole` to the list of things that shouldn't be printed by PrintAST, no?

FWIW, that list of things is here: https://github.com/github/codeql/blob/main/swift/ql/lib/codeql/swift/printast/PrintAstNode.qll#L23

**Review by MathiasVP — COMMENTED** [review]
A couple of suggestions, but otherwise this LGTM!

File: swift/schema.py [hunk 3]
```diff
@@ -1361,3 +1361,31 @@ class BorrowExpr(IdentityExpr):
...
     """
     pass
+
+@qltest.test_with('MacroDecl')
+class MacroRole(AstNode):
+    """
+    The role of a macro, for example #freestanding(declaration) or @attached(member).
+    """
+    kind: int | doc("kind of this macro role (declaration, expression, member, etc.)")
+    macro_syntax: int | doc("#freestanding or @attached")
```
Comment (redsun82) [inline swift/schema.py:1371]: these should probably be marked `| ql.internal`: we generally don't want to expose magic enumeration numbers, and we usually hide them either behind predicates (for example `isDeclarationMacro` etc and `isAttached/isFreestanding`)

File: swift/downgrades/58e9efb426b5b2b322a64c9dab988b02578e0f04/downgrade.ql [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Element extends @element {
+  string toString() { none() }
+}
+
+query predicate new_unspecified_elements(Element e, string property, string error) {
+  unspecified_elements(e, property, error)
+  or
+  error = "Macro declarations removed during database downgrade. Please update your CodeQL code." and
+  property = "" and
+  macro_decls(e, _)
```
Comment (redsun82) [inline swift/downgrades/58e9efb426b5b2b322a64c9dab988b02578e0f04/downgrade.ql:10]: aren't `macro_roles` missing here?

Comment (AlexDenisov) [reply swift/downgrades/58e9efb426b5b2b322a64c9dab988b02578e0f04/downgrade.ql:10]: I think they can be safely deleted as they are only referenced by macro decls?

File: swift/schema.py (hunk 3 above)
Comment (AlexDenisov) [reply swift/schema.py:1371]: I don't mind either way. Should I hide these and add predicates for all the cases?

Comment (redsun82) [reply swift/schema.py:1371]: I'd say yes, @MathiasVP maybe you have something to say?

File: swift/downgrades/58e9efb426b5b2b322a64c9dab988b02578e0f04/downgrade.ql (hunk 4 above)
Comment (redsun82) [reply swift/downgrades/58e9efb426b5b2b322a64c9dab988b02578e0f04/downgrade.ql:10]: you might be right actually

File: swift/schema.py (hunk 3 above)
Comment (MathiasVP) [reply swift/schema.py:1371]: I agree with @redsun82 👍

Comment (AlexDenisov) [reply swift/schema.py:1371]: @redsun82 adding `ql.internal` doesn't seem to have any effect on the generated files, and I don't see it used anywhere except for top-level decls in the schema file.

Comment (redsun82) [reply swift/schema.py:1371]: ah, indeed, I thought I had that already implemented, but I was wrong

Comment (redsun82) [reply swift/schema.py:1371]: I've implemented it in https://github.com/github/codeql/pull/14902 but feel free to ignore it for this PR :slightly_smiling_face:

**Review by redsun82 — APPROVED** [review]
LGTM! Sorry again for the code generation bug.

**Pull request merged**